### PR TITLE
[#2] Setup testing and CI

### LINKED ISSUE #2: Setup testing and CI
## Why

Before other feature tickets come up, we have to set up the CI workflow first.

## Acceptance Criteria

- Setup some simple tests
- Setup CI on Github Action

### PR DESCRIPTION
**Pull request by Lahphim**
* Resolve [Setup testing and CI #2](https://github.com/Lahphim/react-surveyor-challenge/issues/2)

## What happened 👀

Before other feature tickets come up, we have to set up the CI workflow first.
 
## Insight 📝

- Setup Jest for Unit testing
- Setup Cypress for UI testing
- Setup linter
- Add test workflow in GitHub action
 
## Proof Of Work 📹

| CI on Github Action |
| ------------------- |
| Screen Shot 2565-05-12 at 18 32 41 |

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,6 +7,8 @@
 
 # testing
 /coverage
+/cypress/screenshots/*/*.*
+/cypress/videos/*/*.*
```
Comment (carryall) [inline .gitignore:11]: Should we also ignore the `.env` file?

File: cypress/integration/Index/index.spec.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/// <reference types="cypress" />
```
Comment (carryall) [inline cypress/integration/Index/index.spec.ts:1]: I guess we can remove the `.keep` file on the same level as we already have this `Index` folder

Comment (Lahphim) [reply cypress/integration/Index/index.spec.ts:1]: Removed `.keep` https://github.com/Lahphim/react-surveyor-challenge/pull/47/commits/ce5a176e03fbdf0ea70a9eb1cd1c0a885920e4d7

File: .gitignore (hunk 1 above)
Comment (Lahphim) [reply .gitignore:11]: Ignored `.env` file https://github.com/Lahphim/react-surveyor-challenge/pull/47/commits/98ebd0f2f25c2f74bb0c409300d41710dacda14c

File: cypress.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "baseUrl": "http://localhost:3001",
+  "testFiles": ["**/*.spec.ts"]
```
Comment (carryall) [inline cypress.json]: You might want to add a configuration to retry the test if it fails 💡 

```suggestion
  "testFiles": ["**/*.spec.ts"],
  "retries": {
      "runMode": 3,
      "openMode": 2
    }
```

**Review by carryall — APPROVED** [review]

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+          cache: 'yarn'
+
+      - name: Install dependencies
+        run: yarn
+
+      - name: Run linter
+        run: yarn lint
+
+  test-ui:
+    needs: [linter]
```
Comment (hoangmirs) [inline .github/workflows/test.yml]: Why do we need the `linter` step here?
Should we move linter into a dedicated workflow instead? 🤔 
Then we would have `test.yml` and `lint.yml`

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        uses: actions/setup-node@v3
+        with:
+          node-version: '16.15.0'
+          cache: 'yarn'
+
+      - name: Install dependencies
+        run: yarn
+
+      - name: Run linter
+        run: yarn lint
```
Comment (hoangmirs) [inline .github/workflows/test.yml]: Do you have a plan to support `stylelint` in this project? Or just only `eslint`?

Comment (Lahphim) [reply .github/workflows/test.yml]: @hoangmirs I plan to add later on when working with UI setting up

File: .github/workflows/linter.yml [hunk 6]
```diff
@@ -0,0 +1,28 @@
+name: Test
```
Comment (hoangmirs) [inline .github/workflows/linter.yml]: Should it be `Lint` here? And the workflow's name & file name should be a verb following our convention 😄

File: .github/workflows/test.yml (hunk 4 above)
Comment (Lahphim) [reply .github/workflows/test.yml]: Make it in separate https://github.com/Lahphim/react-surveyor-challenge/pull/47/commits/4bf7bdaede665c34b413afd637dcb249da7f3505

File: cypress.json (hunk 3 above)
Comment (Lahphim) [reply cypress.json]: Added retry mode https://github.com/Lahphim/react-surveyor-challenge/pull/47/commits/3d79d932966ae768dbf9450c2a220af575052d36

File: .gitignore [hunk 7]
```diff
@@ -26,6 +28,7 @@ yarn-error.log*
 .pnpm-debug.log*
 
 # local env files
+.env
```
Comment (hoangmirs) [inline .gitignore]: We can use `.env.local` file instead 😄

Comment (Lahphim) [reply .gitignore]: Removed https://github.com/Lahphim/react-surveyor-challenge/pull/47/commits/638716af51845a082db40052d157dc9c56c08ecd

File: .github/workflows/linter.yml (hunk 6 above)
Comment (Lahphim) [reply .github/workflows/linter.yml]: Updated https://github.com/Lahphim/react-surveyor-challenge/pull/47/commits/56ef115b264e6bec6b973876dfcf3acb387db107

**Review by hoangmirs — APPROVED** [review]
LGTM

**Comment by Lahphim** [comment]
Closes #47

**Pull request merged**